materials_io/electron_microscopy.py

```diff
@@ -1,127 +1,1373 @@
-import hyperspy.api as hs
+import logging
+import re
+from typing import Tuple, Dict, Optional
+
+from hyperspy.io import load as hs_load
+from traits.trait_base import Undefined
 
 from materials_io.base import BaseSingleFileParser
+from materials_io.utils import get_nested_dict_value_by_path as get_val
+from materials_io.utils import \
+    map_dict_values, MappingElements, standardize_unit
+from materials_io.utils import set_nested_dict_value_with_units as set_val_units
+
+logger = logging.getLogger(__name__)
 
 
 class ElectronMicroscopyParser(BaseSingleFileParser):
-    """Parse metadata specific to electron microscopy"""
+    """Parse metadata specific to electron microscopy, meaning any file
+    supported by HyperSpy's I/O capabilities. Extract both the metadata
+    interpreted by HyperSpy directly, but also any important values we can
+    pick out manually.
+
+    For each value (if known), return the subdict:
+      {
+        "value": numeric/str value,
+        "unit": unit name from http://qudt.org/vocab/unit/ (see 
+                http://www.qudt.org/doc/DOC_VOCAB-UNITS.html for details)
+      }
+    
+    For base EM, return the following under "General_EM"
+      - accelerating_voltage
+      - acquisition_mode
+      - acquisition_software_name
+      - acquisition_software_version
+      - beam_current (measured at the sample)
+      - beam_energy 
+      - convergence_angle
+      - detector_name
+      - dwell_time (for STEM modes)
+      - emission_current
+      - exposure_time (for non-STEM modes)
+      - elements (list of str - as detected from spectroscopy signal)
+      - frame_time
+      - magnification_actual
+      - magnification_indicated
+      - microscope_name
+      - probe_area
+      - stage_position (dict with X, Y, Z, R, etc. as required)
+
+    For all, return the following under "General" (if known, from HyperSpy,
```

Comment:
Would you be able to express these data format definitions as a JSON schema? We haven't done this with many extractors, but were considering using JSON schemas rather than text as a route for communicating data format. 

We'd probably still want to keep the "what to expect for TEM vs SEM vs ..." parts of the documentation, regardless, but condensing the detailed type information to a schema would make the documentation easier to digest.

If that's too burdensome, let me know.

Reply:
I can give it a shot. I haven't worked with JSON Schema before. I suppose it would go here? https://github.com/materials-data-facility/MaterialsIO/tree/master/materials_io/schemas

Reply:
That's right. As you can see, that's not a feature we've used often. 

Let me know if you want help with the JSON schemas. I've done a lot of work with them, and so has @blaiszik et al

Reply:
JSON Schema isn't too bad, and I think I've got it mostly sorted out. One question... it looks like the schema is not used internally by the parsers at all. Is the intention that after doing a parsing operation, the result would be validated against the schema (if present), or did you not want to put that level of specificity in the code itself?

Reply:
We don't specifically have checking built-in, but using it to double-check the extractor is building correct metadata.

So, if anything, the schemas are more for documentation than quality control.

I'll put in an feature request around adding support for checking